jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp

```diff
@@ -909,57 +947,64 @@ ecma_builtin_global_object_decode_uri_helper (ecma_value_t uri __attr_unused___,
       }
     }
 
-    JERRY_ASSERT (output_start_p + output_size == output_char_p);
-
-    bool valid_utf8 = lit_is_utf8_string_valid (output_start_p, output_size);
-
-    if (valid_utf8)
+    if (ecma_is_completion_value_empty (ret_value))
     {
-      lit_utf8_iterator_t characters = lit_utf8_iterator_create (output_start_p, output_size);
-      output_type_p = output_start_p + output_size;
+      JERRY_ASSERT (output_start_p + output_size == output_char_p);
 
-      while (!lit_utf8_iterator_is_eos (&characters))
-      {
-        bool original_byte = output_type_p[characters.buf_pos.offset] == URI_DECODE_ORIGINAL_BYTE;
+      bool valid_cesu8 = lit_is_cesu8_string_valid (output_start_p, output_size);
 
-        ecma_char_t character = lit_utf8_iterator_read_next (&characters);
+      if (valid_cesu8)
```

Comment:
This complicated post processing is not needed here. The input is valid cesu8,, and invalid parts can only be constructed from %xx parts. Check the 3 octet long %xx sequences, whether they are surrogate pair or not. That is all.